Makefile

```diff
@@ -43,12 +43,24 @@ build-docs:
 docs: build-docs
 	open docs/_build/html/index.html
 
+validate-docs:
```

Comment:
One thing that is missing from this PR (and which is important) is to call this validation in CI. You want that to ensure people don't accidentally commit illegal files which might be hard to spot in a review (e.g. `4711.docs.rst`, `4711.fix.rst`, `4711.feat.rst` all *look*  ok to my tidy reviewer eyes but they are all wrong).   

Looking into the `tox.ini` shows that there's currently a `doctest` job.

https://github.com/ethereum/web3.py/blob/9ec8556ce48d52c858e09e2e2acd42bd71854cfb/tox.ini#L40

Maybe follow what we did with Trinity and turn that into a full `docs` job so that it combines all of this:

1. Build the docs 
2. Validate the doctest snippets
3. Validate the newsfragment files
4. Validate the release notes

There's one extra thing to point out here: `towncrier --draft` is to catch errors in the release notes but is has proven to be a weak solution so far. It doesn't catch actual errors in the RST. So what might happen is that someone adds a newsfragment file and missing a closing backtick or something like that.

We haven't got around fixing that yet but I believe the best case scenario would be to make the doc job be run like this:

1. Validate the doctest snippets
2. Run towncrier using a pseudo version (no `--draft`, right the actual files to disk)
3. Now build the docs
4. Validate the doctest snippets

Using that order, it would be guaranteed that any formatting errors in the release notes are caught by building the docs **after** the release notes where written to disk.

The one downside (which is discussed [here](https://github.com/ethereum/lahja/pull/153#issuecomment-521342852)) is that this command would leave the repository in a dirty state (modified release notes, deleted newsfragments). It's not a big deal in the CI run as nobody cares about the state of the files on a short-lived CI server disk but if someone would run the command locally that might be unexpected. 

To deal with that I'd recommend to do the following:

1. Do not allow the command to run when the repository is in a dirty state
2. Run a `git reset --hard` at the very end to not leave the repository in a dirty state

I'm not saying this PR needs to go that extra mile to do this form of validation, just pointing out what I believe is the best case scenario.

I do think so that the PR needs to ensure the basic validation is triggered in CI.

Reply:
Yeah, I noticed that it wasn't being run in CI and tinkered with it for a little bit yesterday, and then gave up. I was running into an error where CircleCI/pip doesn't know what to do with a project that has both a pyproject.toml and a setup.py. I figured that since I am pretty much the only one releasing, I would just run the `validate-docs` command locally before I released, but I'll try again and hopefully get something auto working.

I don't love the idea of a surprise `git reset --hard`, but would be okay with it in the sequence you described. I'll timebox it today and see if I can get the optimal solution working, and if not I'll leave it for later.